core/src/server/helpers.rs

```diff
@@ -358,25 +494,101 @@ impl BatchResponseBuilder {
 	}
 
 	/// Finish the batch response
-	pub fn finish(mut self) -> String {
+	pub fn finish(mut self) -> BatchResponse {
 		if self.result.len() == 1 {
-			batch_response_error(Id::Null, ErrorObject::from(ErrorCode::InvalidRequest))
+			BatchResponse(batch_response_error(Id::Null, ErrorObject::from(ErrorCode::InvalidRequest)))
 		} else {
 			self.result.pop();
 			self.result.push(']');
-			self.result
+			BatchResponse(self.result)
 		}
 	}
 }
 
+/// Serialized batch response.
+#[derive(Debug, Clone)]
+pub struct BatchResponse(String);
+
 /// Create a JSON-RPC error response.
 pub fn batch_response_error(id: Id, err: impl Into<ErrorObject<'static>>) -> String {
 	let err = ResponsePayload::error_borrowed(err);
 	serde_json::to_string(&Response::new(err, id)).expect("JSON serialization infallible; qed")
 }
 
+/// Similar to [`ResponsePayload`] but possible to with an async-like
+/// API to detect when a method response has been sent.
+#[derive(Debug)]
+pub struct ResponsePayloadV2<'a, T>
```

Comment:
TODO: rename this to `ResponsePayload` technically a breaking change because `ResponsePayload` was enum before but I may treat as non-breaking change to easily backport it v0.20